contrib_bots/lib/giphy/giphy.py

```diff
@@ -0,0 +1,116 @@
+    It also responds to private messages.
+    '''
+    def usage(self):
+        return '''
+            This plugin allows users to post GIFs provided by Giphy.
+            Users should preface keywords with "@giphy" or the Giphy-bot @mention.
+            The bot responds also to private messages.
+            '''
+
+    def get_url_gif_giphy(self, keyword, api_key):
+        # Return a URL for a Giphy GIF based on keywords given.
+        # In case of error, e.g. failure to fetch a GIF URL, it will
+        # return a number.
+        query = {'s': keyword,
+                 'api_key': api_key}
+        try:
+            data = requests.get('http://api.giphy.com/v1/gifs/translate', params=query)
+        except requests.exceptions.ConnectionError as e:  # Usually triggered by bad connection.
+            logging.warning(e)
+            return 1
+
+        if not data.status_code == 200:
+            return 1
+
+        search_status = data.json()['meta']['status']
+        if not search_status == 200:
+            return 1
+
+        try:
+            gif_url = data.json()['data']['images']['original']['url']
+        except TypeError:  # Usually triggered by no result in Giphy.
+            return 2
+
+        return gif_url
+
+    def triage_message(self, message, client):
+        # Return True if we want to (possibly) response to this message.
+        original_content = message['content']
+        is_giphy_called = (original_content.startswith('@giphy ') or
+                           message['is_mentioned'])
+
+        # To prevent infinite loop in private message, bot will detect
+        # if the sender name is the bot name it will return false.
+        if message['type'] == 'private':
+            if client.full_name == message['sender_full_name']:
+                return False
+            else:
+                return True
+
+        return is_giphy_called
+
+    def handle_message(self, message, client, state_handler):
+        new_content = None
+        keyword = None
+
+        # Handle the message that called through mention.
+        if message['is_mentioned']:
+            if message['content'].startswith('@**%s** ' % (client.full_name)):
+                keyword = message['content'][len('@**%s** ' % (client.full_name)):]
+            elif message['content'].startswith('@%s ' % (client.full_name)):
+                keyword = message['content'][len('@%s ' % (client.full_name)):]
+            else:
+                new_content = 'Please mention me first, then type the keyword.'
+        # Handle the message that called through the specified keyword.
+        elif message['content'].startswith('@giphy '):
+            keyword = message['content'][len('@giphy '):]
+        # Handle the private message.
+        elif message['type'] == 'private':
+            keyword = message['content']
+
+        if new_content is None:
+            gif_url = self.get_url_gif_giphy(keyword, config.GIPHY_API_KEY)
```

Comment:
hmm, actually if `keyword` is `None` that means `new_content` is defined like line 77, and if `new_content` is defined, automatically line 86 is not executed